internal/pkg/encoding/fxamacker_cbor.go

```diff
@@ -0,0 +1,192 @@
+package encoding
+
+import (
+	"bytes"
+	"fmt"
+
+	cbor "github.com/fxamacker/cbor"
+	hamt "github.com/ipfs/go-hamt-ipld"
+	cbg "github.com/whyrusleeping/cbor-gen"
+
+	e "github.com/filecoin-project/go-filecoin/internal/pkg/enccid"
+)
+
+// FxamackerCborEncoder is an object encoder that encodes objects based on the CBOR standard.
+type FxamackerCborEncoder struct {
+	b bytes.Buffer
+}
+
+// FxamackerCborDecoder is an object decoder that decodes objects based on the CBOR standard.
+type FxamackerCborDecoder struct {
+	raw []byte
+}
+
+// NewFxamackerCborEncoder creates a new `FxamackerCborEncoder`.
+func NewFxamackerCborEncoder() FxamackerCborEncoder {
+	return FxamackerCborEncoder{}
+}
+
+// NewFxamackerCborDecoder creates a new `FxamackerCborDecoder`.
+func NewFxamackerCborDecoder(b []byte) FxamackerCborDecoder {
+	return FxamackerCborDecoder{
+		raw: b,
+	}
+}
+
+//
+// FxamackerCborEncoder
+//
+
+// EncodeUint encodes a uint.
+func (encoder *FxamackerCborEncoder) EncodeUint(obj uint) error {
+	return encoder.encodeCbor(obj)
+}
+
+// EncodeUint8 encodes a uint8.
+func (encoder *FxamackerCborEncoder) EncodeUint8(obj uint8) error {
+	return encoder.encodeCbor(obj)
+}
+
+// EncodeUint16 encodes a uint16.
+func (encoder *FxamackerCborEncoder) EncodeUint16(obj uint16) error {
+	return encoder.encodeCbor(obj)
+}
+
+// EncodeUint32 encodes a uint32.
+func (encoder *FxamackerCborEncoder) EncodeUint32(obj uint32) error {
+	return encoder.encodeCbor(obj)
+}
+
+// EncodeUint64 encodes a uint64.
+func (encoder *FxamackerCborEncoder) EncodeUint64(obj uint64) error {
+	return encoder.encodeCbor(obj)
+}
+
+// EncodeInt encodes a int.
+func (encoder *FxamackerCborEncoder) EncodeInt(obj int) error {
+	return encoder.encodeCbor(obj)
+}
+
+// EncodeInt8 encodes a int8.
+func (encoder *FxamackerCborEncoder) EncodeInt8(obj int8) error {
+	return encoder.encodeCbor(obj)
+}
+
+// EncodeInt16 encodes a int16.
+func (encoder *FxamackerCborEncoder) EncodeInt16(obj int16) error {
+	return encoder.encodeCbor(obj)
+}
+
+// EncodeInt32 encodes a int32.
+func (encoder *FxamackerCborEncoder) EncodeInt32(obj int32) error {
+	return encoder.encodeCbor(obj)
+}
+
+// EncodeInt64 encodes a int64.
+func (encoder *FxamackerCborEncoder) EncodeInt64(obj int64) error {
+	return encoder.encodeCbor(obj)
+}
+
+// EncodeBool encodes a bool.
+func (encoder *FxamackerCborEncoder) EncodeBool(obj bool) error {
+	return encoder.encodeCbor(obj)
+}
+
+// EncodeString encodes a string.
+func (encoder *FxamackerCborEncoder) EncodeString(obj string) error {
+	return encoder.encodeCbor(obj)
+}
+
+// EncodeArray encodes an array.
+func (encoder *FxamackerCborEncoder) EncodeArray(obj interface{}) error {
+	return encoder.encodeCbor(obj)
+}
+
+// EncodeMap encodes a map.
+func (encoder *FxamackerCborEncoder) EncodeMap(obj interface{}) error {
+	return encoder.encodeCbor(obj)
+}
+
+// EncodeStruct encodes a struct.
+func (encoder *FxamackerCborEncoder) EncodeStruct(obj interface{}) error {
```

Comment:
If the struct implements CBORMarshal, I think we need a really good reason not to use it.